### PR TITLE
Bugs/341 rework mpi communication wrappers

### PR DESCRIPTION
**Pull request by mrfh92**
The following changes have been done: 

* the generic wrapper `__scatter_like` that had been used for `(I)Scatter`, `(I)Scatterv`, `(I)Gather`, `(I)Gatherv` has been splitted up into wrappers `__scatter_like` (for `(I)Scatter`, `(I)Scatterv`) and `__gather_like` (for  `(I)Gather`, `(I)Gatherv`)

* some smaller corrections: in many places axis-permutation of `recv_buf` took place in an if-clause that also depends on `CUDA_AWARE_MPI` resulting in the problem that this permutation took only place if buffers were on GPU but CUDA_AWARE_MPI was disabled... this has been corrected by splitting this into two if-clauses 

* a small change in `printing.py`: in #1121 we had corrected an error using `gather` (fallback from mpi4py) since `Gatherv` did not work. This has been corrected now.

### CODE REVIEW

**Comment by mrfh92** [comment]
outcome of tests on terrabyte (2 nodes, 4 tasks each, 1GPU/2 CPU-cores per task): 
* problem with random on both GPU and CPU 
* problem with DASO on GPU only

**Comment by mrfh92** [comment]
Successfully tested on HDFML with 
* GPU: 2 nodes, 4 tasks per node (with 6 and 2 CPUs per task) 
* GPU: 1 node, 4 tasks per node (2 CPUs per task)
* CPU: 2 nodes, 4 tasks per node, 6 and 2 CPUs per task 
* CPU: 1 node, 4 tasks per node, 2 CPUs per task

**Comment by mrfh92** [comment]
@ClaudiaComito @mtar as far as I can see, this PR should be ready for merging... so feel free to do so even if Im not in the PR meeting next week

**Comment by mtar** [comment]
@mrfh92 Have you switched the base branch when working on it? The first two commits are from release, but then you merge from main.

**Comment by mtar** [comment]
Also note that some lines are not covered.

File: heat/core/version.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 """Indicates feature extension."""
 micro: int = 0
 """Indicates revisions for bugfixes."""
-extension: str = "dev"
+extension: str = None
```
Comment (ClaudiaComito) [inline heat/core/version.py:10]:
```suggestion
extension: str = dev
```

File: heat/core/version.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 """Indicates feature extension."""
 micro: int = 0
 """Indicates revisions for bugfixes."""
-extension: str = "dev"
+extension: str = dev
```
Comment (ClaudiaComito) [inline heat/core/version.py:10]:
```suggestion
extension: str = "dev"
```

**Review by ClaudiaComito — APPROVED** [review]
Looks great, thanks @mrfh92 !

**Pull request merged**